dynamodb/schema.go

```diff
@@ -0,0 +1,358 @@
+	dySchema := dynamoDBSchema{TableName: table}
+	err := dySchema.parseIndexes(client)
+	if err != nil {
+		return err
+	}
+	stats, count, err := dySchema.scanSampleData(client, sampleSize)
+	if err != nil {
+		return err
+	}
+	dySchema.inferDataTypes(stats, count)
+	conv.SrcSchema[table] = dySchema.genericSchema()
+	return nil
+}
+
+type dynamoDBSchema struct {
+	TableName   string
+	ColumnNames []string
+	ColumnTypes map[string]colType
+	PrimaryKeys []string
+	SecIndexes  []index
+}
+
+type colType struct {
+	Name     string
+	Nullable bool
+}
+
+type index struct {
+	Name string
+	Keys []string
+}
+
+func (s *dynamoDBSchema) parseIndexes(client dynamoClient) error {
+	input := &dynamodb.DescribeTableInput{
+		TableName: aws.String(s.TableName),
+	}
+
+	result, err := client.DescribeTable(input)
+	if err != nil {
+		return fmt.Errorf("failed to make a DescribeTable API call for table %v: %v", s.TableName, err)
+	}
+
+	// Primary keys
+	for _, i := range result.Table.KeySchema {
+		s.PrimaryKeys = append(s.PrimaryKeys, *i.AttributeName)
+	}
+
+	// Secondary indexes
+	for _, i := range result.Table.GlobalSecondaryIndexes {
+		var keys []string
+		for _, j := range i.KeySchema {
+			keys = append(keys, *j.AttributeName)
+		}
+		s.SecIndexes = append(s.SecIndexes, index{Name: *i.IndexName, Keys: keys})
+	}
+
+	return nil
+}
+
+func (s *dynamoDBSchema) scanSampleData(client dynamoClient, sampleSize int64) (map[string]map[string]int64, int64, error) {
+	// A map from column name to a count map of possible data types.
+	stats := make(map[string]map[string]int64)
+	var lastEvaluatedKey map[string]*dynamodb.AttributeValue
+	var count int64
+
+	for {
+		// Build the query input parameters
+		params := &dynamodb.ScanInput{
+			TableName: aws.String(s.TableName),
+		}
+		if lastEvaluatedKey != nil {
+			params.ExclusiveStartKey = lastEvaluatedKey
+		}
+
+		// Make the DynamoDB Query API call
+		result, err := client.Scan(params)
+		if err != nil {
+			return nil, 0, fmt.Errorf("failed to make Query API call for table %v: %v", s.TableName, err)
+		}
+
+		// Iterate the items returned
+		for _, attrsMap := range result.Items {
+			for attrName, attr := range attrsMap {
+				if _, ok := stats[attrName]; !ok {
+					stats[attrName] = make(map[string]int64)
+				}
+				incDyDataTypeCount(attrName, attr, stats[attrName])
+			}
+
+			count++
+			if count >= sampleSize {
+				return stats, count, nil
+			}
+		}
+		if result.LastEvaluatedKey == nil {
+			break
+		}
+		// If there are more rows, then continue.
+		lastEvaluatedKey = result.LastEvaluatedKey
+	}
+	return stats, count, nil
+}
+
+func incDyDataTypeCount(attrName string, attr *dynamodb.AttributeValue, s map[string]int64) {
+	if attr.S != nil {
+		incCount(s, typeString)
+	} else if attr.BOOL != nil {
+		incCount(s, typeBool)
+	} else if attr.N != nil {
+		if int64Parsable(*attr.N) {
+			incCount(s, typeNumberInt)
+		} else {
```

Comment:
Maybe check that *attr.N is parseable as a float64?
And then use string if not.

Reply:
I'm not sure whether or not it considers `parsable`, even if the number cannot be represented accurately in float64. 

We also should consider whether or not this number can be converted to a Numeric. I'll give it a try.